DatabaseWorkshop.md

```diff
@@ -0,0 +1,185 @@
+All of the handlers now need to have `@escaping` added to their function signature, as the ORM uses an escaping method. Add `@escaping` to each method, after `completion:`. For example:
+```swift
+func storeHandler(todo: ToDo, completion: @escaping (ToDo?, RequestError?) -> Void ) { ...
+func deleteAllHandler(completion: @escaping (RequestError?) -> Void ) { ...
+```
+### Step 7 - storeHandler() updates
+We will first update storeHandler() to use the database. Remove all of the code between the methods curly braces and paste in the following:
+```swift
+		var todo = todo
+		if todo.completed == nil {
+			todo.completed = false
+		}
+		todo.id = nextId
+		todo.url = "http://localhost:8080/\(nextId)"
+		nextId += 1
+		todo.save(completion)
+```
+This code is similar to the original but uses `todo.save(completion)` which saves the `todo` object to the database and then calls the methods completion handler.
+### Step 8 - deleteOne(), deleteAll(), getOne(), getAll()
+These methods are easy to update as their logic can happen with one call using the ORM! 
+```swift
+	func deleteAllHandler(completion: @escaping (RequestError?) -> Void ) {
+		ToDo.deleteAll(completion)
+	}
+
+	func deleteOneHandler(id: Int, completion: @escaping (RequestError?) -> Void ) {
+		ToDo.delete(id: id, completion)
+	}
+	
+	func getAllHandler(completion: @escaping ([ToDo]?, RequestError?) -> Void ) {
+		ToDo.findAll(completion)
+	}
+	
+	func getOneHandler(id: Int, completion: @escaping(ToDo?, RequestError?) -> Void ) {
+		ToDo.find(id: id, completion)
+	}
+```
+Each one now uses one call to the database to accomplish it's task. The code is also easy to understand just from reading the calls happening inside the methods.
+
+### Step 9 - updateHandler()  
+
+This one has a little more going on as we need to fetch the version of the `ToDo` currently stored in our database, assess the differences between this version of the `ToDo` and the patched version being passed into the method, make the changes and then commit the modified `ToDo` back to the database.
+
+```swift
+	func updateHandler(id: Int, new: ToDo, completion: @escaping (ToDo?, RequestError?) -> Void ) {
+
+		ToDo.find(id: id) { (preExistingToDo, error) in
+			if error != nil {
+				return completion(nil, .notFound)
+			}
+			
+			guard var oldToDo = preExistingToDo else {
+				return completion(nil, .notFound)
+			}
+			
+			guard let id = oldToDo.id else {
+				return completion(nil, .internalServerError)
+			}
+			
+			oldToDo.user = new.user ?? oldToDo.user
+			oldToDo.order = new.order ?? oldToDo.order
+			oldToDo.title = new.title ?? oldToDo.title
+			oldToDo.completed = new.completed ?? oldToDo.completed
+			
+			oldToDo.update(id: id, completion)
+			
+		}
+	}
+```
+
+Again, the logic is similar to before but with some added error handling for fetching from the database incase the `id` doesn't exist. 
+
+The final step is now to remove the `todoStore` variable from the top of the `App` Class.
```

Comment:
I think this should be a separate step e.g. 

10. Remove the `todoStore` variable

The final step is to remove the `todoStore` variable definition from the top of the `App` class within `Application.swift`.

Reply:
✅